src/Google.Api.Gax.Grpc/ApiCall.cs

```diff
@@ -33,10 +33,65 @@ internal static class ApiCall
             where TRequest : class, IMessage<TRequest>
             where TResponse : class, IMessage<TResponse>
         {
-            return new ApiCall<TRequest, TResponse>(
-                (req, cs) => asyncGrpcCall(req, cs.ToCallOptions(clock)).ResponseAsync,
-                (req, cs) => syncGrpcCall(req, cs.ToCallOptions(clock)),
-                baseCallSettings);
+            var adapter = new GrpcCallAdapter<TRequest, TResponse>(asyncGrpcCall, syncGrpcCall, clock);
+            return new ApiCall<TRequest, TResponse>(adapter.CallAsync, adapter.CallSync, baseCallSettings);
+        }
+
+        private class GrpcCallAdapter<TRequest, TResponse>
+            where TRequest : class, IMessage<TRequest>
+            where TResponse : class, IMessage<TResponse>
+        {
+            private readonly Func<TRequest, CallOptions, AsyncUnaryCall<TResponse>> _asyncGrpcCall;
+            private readonly Func<TRequest, CallOptions, TResponse> _syncGrpcCall;
+            private readonly IClock _clock;
+
+            internal GrpcCallAdapter(
+                Func<TRequest, CallOptions, AsyncUnaryCall<TResponse>> asyncGrpcCall,
+                Func<TRequest, CallOptions, TResponse> syncGrpcCall,
+                IClock clock)
+            {
+                _asyncGrpcCall = asyncGrpcCall;
+                _syncGrpcCall = syncGrpcCall;
+                _clock = clock;
+            }
+
+            internal Task<TResponse> CallAsync(TRequest request, CallSettings callSettings)
+            {
+                var unaryCall = _asyncGrpcCall(request, callSettings.ToCallOptions(_clock));
+                var responseMetadataHandler = callSettings.ResponseMetadataHandler;
+                var trailingMetadataHandler = callSettings.TrailingMetadataHandler;
+
+                // If we don't have any response/trailing metadata handlers, it's simplest just to
+                // return the gRPC task.
+                if (responseMetadataHandler == null && trailingMetadataHandler == null)
+                {
+                    return unaryCall.ResponseAsync;
+                }
+
+                // Otherwise, we want to pass the returned metadata to the handlers before our returned task completes,
+                // so we create a proxy task.
+                return unaryCall.ResponseAsync.ContinueWith(task =>
+                {
+                    var responseHeadersTask = unaryCall.ResponseHeadersAsync;
+                    if (responseHeadersTask.Status == TaskStatus.RanToCompletion)
+                    {
+                        responseMetadataHandler?.Invoke(responseHeadersTask.Result);
+                    }
+                    trailingMetadataHandler?.Invoke(unaryCall.GetTrailers());
```

Comment:
Handlers are called synchronously, so they have to complete before the RPC is complete.
I agree this is the simplest thing to do.